pkg/controller/common.go

```diff
@@ -68,17 +86,323 @@ type CloudProvisioner interface {
+	}
+	claims, ok = value.([]string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "wrong output type: expected []string")
+	}
+
+	volumes := []string{}
+	for _, claim := range claims {
+		value, ok := c.claimToVolumeMap.Load(claim)
+		if !ok {
+			// the pvc entry is not an azure resource
+			continue
+		}
+		volume, ok := value.(string)
+		if !ok {
+			return nil, status.Errorf(codes.Internal, "wrong output type: expected string")
+		}
+		volumes = append(volumes, volume)
+	}
+	return volumes, nil
+}
+
+func (c *SharedState) getPodsFromVolume(volumeName string) ([]string, error) {
+	v, ok := c.volumeToClaimMap.Load(volumeName)
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "no bound persistent volume claim was found for AzVolume (%s)", volumeName)
+	}
+	claimName, ok := v.(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "volumeToClaimMap should should hold string")
+	}
+
+	value, ok := c.claimToPodsMap.Load(claimName)
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "no pods found for PVC (%s)", claimName)
+	}
+	pods, ok := value.([]string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "claimToPodsMap should hold string slices")
+	}
+	return pods, nil
+}
+
+func (c *SharedState) getNodesForReplica(ctx context.Context, azr azReconciler, volumes []string, pods ...string) ([]string, error) {
+	if volumes == nil {
+		volumes = []string{}
+		for _, pod := range pods {
+			podVolumes, err := c.getVolumesFromPod(pod)
+			if err != nil {
+				return nil, err
+			}
+			volumes = append(volumes, podVolumes...)
+		}
+	}
+
+	numReplica := 0
+	for _, volume := range volumes {
+		var azVolume v1alpha1.AzVolume
+		if err := azr.getClient().Get(ctx, types.NamespacedName{Namespace: azr.getNamespace(), Name: volume}, &azVolume); err != nil {
+			return nil, err
+		}
+		numReplica = max(numReplica, azVolume.Spec.MaxMountReplicaCount)
+	}
+
+	nodes := []string{}
+	nodeScores := map[string]int{}
+	primaryNode := ""
+	for _, volume := range volumes {
+		volReq, err := createLabelRequirements(azureutils.VolumeNameLabel, volume)
+		if err != nil {
+			return nil, err
+		}
+		labelSelector := labels.NewSelector().Add(*volReq)
+		azVolumeAttachments := &v1alpha1.AzVolumeAttachmentList{}
+		err = azr.getClient().List(ctx, azVolumeAttachments, &client.ListOptions{LabelSelector: labelSelector})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, azVolumeAttachment := range azVolumeAttachments.Items {
+			if azVolumeAttachment.Spec.RequestedRole == v1alpha1.PrimaryRole {
+				if primaryNode == "" {
+					primaryNode = azVolumeAttachment.Spec.NodeName
+				}
+				continue
+			}
+			if _, ok := nodeScores[azVolumeAttachment.Spec.NodeName]; !ok {
+				nodeScores[azVolumeAttachment.Spec.NodeName] = 0
+				nodes = append(nodes, azVolumeAttachment.Spec.NodeName)
+			}
+			nodeScores[azVolumeAttachment.Spec.NodeName]++
+		}
+	}
+	sort.Slice(nodes[:], func(i, j int) bool {
```

Comment:
Since, we are sorting by the nodes that have maximum number of attachments, is there a way to check that that the node hasn't already exhausted its attachment limit? If that's the case, we should probably not use that node.

Reply:
I will be patching this function along with `selectNodes` that puts node capacity and fault domain into account in later PRs. :)